### PR TITLE
feat: support ascend npu sft

### PR DESCRIPTION
**Pull request by donpromax**
## Summary

This PR adds Ascend/NPU support for SFT workloads in StepTronOSS.

It introduces a NPU runtime patch entrypoint, NPU-specific optimization backends for MoE dispatch and grouped GEMM, Ascend SFT experiment entrypoints,
benchmark scripts, tests, and user-facing documentation.

## Main Changes

- add manual NPU runtime patching through `steptronoss.utils.npu_patch.apply_npu_patch()`
- add `grouped_gemm="npu_gmm"` backed by MindSpeed `npu_gmm_v2`
- add `TokenDispatcher="npu_alltoall"` for EP token routing on NPU
- patch several CUDA-specific helper paths to work on NPU runtime
- register NPU backends when `apply_npu_patch()` is called
- add Ascend SFT experiment entrypoints for:
  - Qwen3-1.7B
  - Qwen3-30A3B
  - Step3 toy SFT
  - Step3.5-Flash Midtrain SFT with Muon
- add a real-data preparation script for Qwen3 SFT smoke and validation runs
- add NPU benchmarks and test cases for dispatcher and grouped GEMM
- add user documentation:
  - `docs/ASCEND.md`
  - `docs/ASCEND_ZH.md`

## Implementation Notes

- Ascend patching is now explicitly enabled by calling `apply_npu_patch()` before importing modules that depend on NPU runtime behavior or before selecting NPU backends
- `npu_alltoall` builds a unique token-rank routing layout before communication to avoid duplicated routing to the same remote rank
- the fast path uses MindSpeed `npu_moe_token_permute` / `npu_moe_token_unpermute` on the NPU bf16 path
- the dispatcher keeps an `index_select` / `index_add` fallback path for compatibility
- `npu_gmm` keeps the same semantic interface as the existing `grouped_gemm` path and normalizes `batch_sizes` internally

## Tested Environment

Validated with:

- CANN `8.3.RC2`
- `torch_npu` `2.8`
- MindSpeed `v2.2.0_core_r0.12.1`

All other dependencies are kept aligned with the project's `uv` environment.

## Validation

- added dispatcher unit tests in `tests/test_npu_alltoall_dispatcher.py`
- added grouped GEMM correctness test in `tests/test_grouped_gemm_npu.py`
- added microbenchmarks:
  - `benchmarks/benchmark_dispatcher_npu.py`
  - `benchmarks/benchmark_grouped_gemm_npu.py`
- ran end-to-end SFT validation on representative Qwen3 and Step3.5 configs

## Training Results

Qwen3-1.7B sft smoke test (compared to l40s):
910b_compare_to_l40s

Step-3.5-Flash-Base-Midtrain sft: tp8pp8vpp3ep8 seq_len=8192 with recompute, sequence_parallel and offload_optimizer_state
image

## Kernel Optimization

NPU grouped GEMM speedup vs baseline:
```text
[npu_grouped_gemm] name=moe_like_large group=36 batch=3256 k=4096 n=2560 dtype=bf16 trans_b=True
backend, fw_ms, bw_ms, total_ms
baseline, 11.976, 235.589, 247.565
npu_gmm_v2, 8.257, 19.858, 28.115
speedup_vs_baseline, fw=1.45x, bw=11.86x, total=8.81x
metric, close, max_abs_diff
forward, True, 0.000000
forward_bw_run, True, 0.000000
grad_a, True, 0.000000
grad_b, True, 0.000000
```

NPU MoE AllToAll Dispatcher speedup vs baseline:
dispatcher_npu_seq_len_speedup_20260325 (1)

### CODE REVIEW

File: benchmarks/benchmark_dispatcher_npu.py [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+import os
+import sys
+import time
+import threading
+from pathlib import Path
+
+import torch
+import torch.distributed as dist
+
+try:
```
Comment (Randomizez) [inline benchmarks/benchmark_dispatcher_npu.py]: 这个可以直接加到项目依赖，不用这么处理

File: benchmarks/benchmark_dispatcher_npu.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+    import psutil
+except Exception:
+    psutil = None
+
+REPO_ROOT = Path(__file__).resolve().parent
```
Comment (Randomizez) [inline benchmarks/benchmark_dispatcher_npu.py:21]: 这是为什么？

File: steptronoss/utils/npu_patch.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+            val_typecheck = item.data
+        # NOTE: NPU fp16/bf16 tensors skip the original CUDA-only type check.
+        if torch.is_tensor(val_typecheck) and val_typecheck.dtype in (torch.float16, torch.bfloat16):
+            item = item.float()
+        return item
+
+    return conversion_helper(val, float_conversion)
+
+
+def _register_npu_grouped_gemm_optimization():
```
Comment (Randomizez) [inline steptronoss/utils/npu_patch.py]: 这个能否直接写到grouped_gemm的alternatives里？用lazy import应该就可以。这个patch有点脏

File: steptronoss/utils/npu_patch.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+    register = OPTIMIZABLE_REGISTER.get(TOKEN_DISPATCHER_REGISTER_NAME)
+    if register is None:
+        logger.warning(f"Cannot find optimizable register entry: {TOKEN_DISPATCHER_REGISTER_NAME}")
+        return
+    register["alternatives"]["npu_alltoall"] = NPUAllToAllDispatcher
+
+
+def _apply_steptron_ascend_patches():
+    logger.info("Applying StepTron Ascend patches.", at=0)
+    StepTronPatchesManager.register_patch(
```
Comment (Randomizez) [inline steptronoss/utils/npu_patch.py]: 这里能否改为AttentionCore的一个alternative，现在使用ascend的路径是sdpa -> alt flash-attn -> patched forward，有点绕。直接sdpa -> alt npu_sdpa行吗

File: steptronoss/utils/npu_patch.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        force_patch=True,
+    )
+    StepTronPatchesManager.apply_patches()
+    _register_npu_grouped_gemm_optimization()
+    _register_npu_alltoall_dispatcher_optimization()
+
+
+def apply_npu_patch():
+    global NPU_PATCH_ACTIVE
+    if not NPU_PATCH_ACTIVE and _torch_npu_available():
```
Comment (Randomizez) [inline steptronoss/utils/npu_patch.py]: 这里else可以直接raise，npu not available直接报错避免非预期行为

File: playground/sft/step3/npu/step3p5_flash_sft_step3_data_muon_npu.py
Comment (Randomizez) [inline playground/sft/step3/npu/step3p5_flash_sft_step3_data_muon_npu.py:1]: 这里是不是本质上只需要apply_patch然后set_optimization即可？能否写成来表达最少需要多少替换？
```
apply_patch()
from xxx import Exp as BaseExp

class Exp(BaseExp):
    def set_optimization(self):
        # use ascend alternatives
```

**Review by Randomizez — COMMENTED** [review]
感觉有很多可以简化的地方，请看一下。另外alternatives尽量直接注册，在patch内注册有点太过隐式了。

File: benchmarks/benchmark_dispatcher_npu.py (hunk 2 above)
Comment (donpromax) [reply benchmarks/benchmark_dispatcher_npu.py:21]: benchmarks/benchmark_grouped_gemm.py 其他的bench好像也带了这一段，可能是为了没有pip install -e .也可以运行测试？

File: steptronoss/utils/npu_patch.py (hunk 3 above)
Comment (donpromax) [reply steptronoss/utils/npu_patch.py]: Done

File: steptronoss/utils/npu_patch.py (hunk 5 above)
Comment (donpromax) [reply steptronoss/utils/npu_patch.py]: Done

File: playground/sft/step3/npu/step3p5_flash_sft_step3_data_muon_npu.py
Comment (donpromax) [reply playground/sft/step3/npu/step3p5_flash_sft_step3_data_muon_npu.py:1]: 是的，但是这个文件还改了一些参数，确保可以在65G * 64卡可以不OOM

**Review by Randomizez — APPROVED** [review]
LGTM
